mce-hbtimer.c

```diff
@@ -0,0 +1,1095 @@
+    int         hbt_period;
+
+    /** Flag for: control within hbt_notify() */
+    bool        hbt_in_notify;
+
+    /** User data to pass to hbt_notify() */
+    void       *hbt_user_data;
+};
+
+mce_hbtimer_t * mce_hbtimer_create         (const char *name, int period, GSourceFunc notify, void *user_data);
+void            mce_hbtimer_delete         (mce_hbtimer_t *self);
+
+bool            mce_hbtimer_is_active      (const mce_hbtimer_t *self);
+const char     *mce_hbtimer_get_name       (const mce_hbtimer_t *self);
+void            mce_hbtimer_set_period     (mce_hbtimer_t *self, int period);
+void            mce_hbtimer_start          (mce_hbtimer_t *self);
+void            mce_hbtimer_stop           (mce_hbtimer_t *self);
+
+static void     mce_hbtimer_notify         (mce_hbtimer_t *self);
+static void     mce_hbtimer_set_trigger    (mce_hbtimer_t *self, int64_t trigger);
+
+/* ------------------------------------------------------------------------- *
+ * GENERIC_UTILITIES
+ * ------------------------------------------------------------------------- */
+
+/** Monotonic tick value used to signify "not-set" */
+#define NO_TICK INT64_MAX
+
+static int64_t  mht_get_monotick           (void);
+static guint    mht_add_iowatch            (int fd, bool close_on_unref, GIOCondition cnd, GIOFunc io_cb, gpointer aptr);
+
+/* ------------------------------------------------------------------------- *
+ * QUEUE_MANAGEMENT
+ * ------------------------------------------------------------------------- */
+
+/** List of registered timers */
+static GSList *mht_queue_timer_list = 0;
+
+void            mht_queue_dispatch_timers  (void);
+static void     mht_queue_schedule_wakeups (void);
+static void     mht_queue_garbage_collect  (void);
+static void     mht_queue_add_timer        (mce_hbtimer_t *self);
+static void     mht_queue_remove_timer     (mce_hbtimer_t *self);
+static bool     mht_queue_has_timer        (const mce_hbtimer_t *self);
+
+/* ------------------------------------------------------------------------- *
+ * GLIB_WAKEUPS
+ * ------------------------------------------------------------------------- */
+
+/** Source ID for currently active glib timer wakeup */
+static guint mce_hbtimer_glib_wait_id = 0;
+
+static gboolean mht_glib_wakeup_cb         (gpointer aptr);
+static void     mht_glib_set_wakeup        (int64_t trigger, int64_t now);
+
+/* ------------------------------------------------------------------------- *
+ * IPHB_WAKEUPS
+ * ------------------------------------------------------------------------- */
+
+/** How much wakeups from suspend can be delayed [s]
+ *
+ * To increase changes of aligning with other iphb wakeups this
+ * should be >= heartbeat period (=12 seconds) */
+#define MHT_IPHB_WAKEUP_MAX_DELAY_S     12
+
+/** Cached timestamp of last requested iphb wakeup */
+static int64_t  mht_iphb_wakeup_tick = NO_TICK;
+
+/** Source id for iphb wakeup input watch */
+static guint   zzz_hb_wakeup_watch_id = 0;
```

Comment:
I have a habit of initially using easy-to-search-and-replace prefixes like "zzz_" and "qqq_" until I'm fairly certain that I know what it is going to be for real... looks like I left some of those in place.

Will fix.